sources/pg_replication/helpers.py

```diff
@@ -0,0 +1,757 @@
+    options = {"publication_names": pub_name, "proto_version": "1"}
+    upto_lsn = get_max_lsn(slot_name, options, credentials)
+    if upto_lsn is None:
+        return "Replication slot is empty."
+
+    # generate items in batches
+    while True:
+        gen = ItemGenerator(
+            credentials=credentials,
+            slot_name=slot_name,
+            options=options,
+            upto_lsn=upto_lsn,
+            start_lsn=start_lsn,
+            target_batch_size=target_batch_size,
+            include_columns=include_columns,
+            columns=columns,
+            write_disposition=write_disposition,  # TODO: remove after https://github.com/dlt-hub/dlt/issues/1031 has been released
+        )
+        yield from gen
+        if gen.generated_all:
+            dlt.current.resource_state(resource_name)[
+                "last_commit_lsn"
+            ] = gen.last_commit_lsn
+            break
+        start_lsn = gen.last_commit_lsn
+
+
+def _get_conn(
+    credentials: ConnectionStringCredentials,
+    connection_factory: Optional[Any] = None,
+) -> Union[psycopg2.extensions.connection, LogicalReplicationConnection]:
+    """Returns a psycopg2 connection to interact with postgres."""
+    return psycopg2.connect(  # type: ignore[no-any-return]
+        database=credentials.database,
+        user=credentials.username,
+        password=credentials.password,
+        host=credentials.host,
+        port=credentials.port,
+        connection_factory=connection_factory,
+    )
+
+
+def _get_rep_conn(
+    credentials: ConnectionStringCredentials,
+) -> LogicalReplicationConnection:
+    """Returns a psycopg2 LogicalReplicationConnection to interact with postgres replication functionality.
+
+    Raises error if the user does not have the REPLICATION attribute assigned.
+    """
+    return _get_conn(credentials, LogicalReplicationConnection)  # type: ignore[return-value]
+
+
+def _make_qualified_table_name(table_name: str, schema_name: str) -> str:
+    """Escapes and combines a schema and table name."""
+    return (
+        escape_postgres_identifier(schema_name)
+        + "."
+        + escape_postgres_identifier(table_name)
+    )
+
+
+def _gen_replication_resource_name(slot_name: str, pub_name: str) -> str:
+    """Generates name for a resource used for replication.
+
+    Based on names of replication slot and publication the resource consumes from.
+    """
+    return slot_name + "_" + pub_name
+
+
+def _get_pk(
+    table_name: str,
+    schema_name: str,
+    credentials: ConnectionStringCredentials,
+) -> Optional[TColumnNames]:
+    """Returns primary key column(s) for postgres table.
+
+    Returns None if no primary key columns exist.
+    """
+    qual_name = _make_qualified_table_name(table_name, schema_name)
+    cur = _get_conn(credentials).cursor()
+    # https://wiki.postgresql.org/wiki/Retrieve_primary_key_columns
+    cur.execute(
+        f"""
+        SELECT a.attname
```

Comment:
The `primary_key` hint on the `sql_table` resource gives an empty list, so perhaps something's going wrong there:
![image](https://github.com/dlt-hub/verified-sources/assets/47451109/167040f6-ed9c-4a20-90a2-97dcbd3d50b3)

Reply:
Okay I get it now. `sql_table` creates a resource for the snapshot table, which doesn't have a primary key, because it is created with a CTAS statement:

```python
  cur.execute(
      f"""
      START TRANSACTION ISOLATION LEVEL REPEATABLE READ;
      SET TRANSACTION SNAPSHOT '{snapshot_name}';
      CREATE TABLE {snapshot_qual_name} AS SELECT {col_str} FROM {qual_name};
  """
  )
```

I point `_get_pk` at the original table (`table_name` instead of `snapshot_table_name`), which does have a primary key, hence the difference.